### PR TITLE
Reworking and Consolidation of the Base Space ROS Image Build

### PR DESCRIPTION
**Pull request by eholum**
Resolves space-ros/space-ros#176 and space-ros/space-ros#175.

Based on:

* https://github.com/space-ros/space-ros/pull/164
* https://github.com/space-ros/space-ros/pull/167

This PR:

* Consolidates the repos file generation and build Earthfiles
* Removes a bunch of the build logic around branches of this repo
* Moves the spaceros install to /opt/spaceros from the created users home directory
  * I think it makes much more sense to put it there, as it's in line with the general ros approach
* Includes some minor optimizations for the build
* General cleanup of the documentation
* Keeps spaceros scripts in /opt/spaceros/scripts

In searching through the overlaid applications and demonstrations in `docker`, I haven't found functional references to the`spaceros` install in the user's home directory. Though there are some docs references that will have to be updated alongside of this.

This does not yet separate dev/base targets, which is next on the todo list.

### CODE REVIEW

**Comment by eholum** [comment]
@Bckempa if you have some time to take a scan through this it would be much appreciated.

If this goes in we will need a follow on task in the docker repo to remove README references to the spaceros install in the user's home directory. This might be the only one but I will take a closer look, too: https://github.com/space-ros/docker/blob/main/space_robots/README.md#setup

**Comment by asimonov** [comment]
> Resolves #176.
> 
> Based on:
> 
>     * [Parameterize the ros2.repos generation script #164](https://github.com/space-ros/space-ros/pull/164)
> 
>     * [Redo: Move spaceros Base Image from docker to space-ros #167](https://github.com/space-ros/space-ros/pull/167)
> 
> 
> This PR:
> 
>     * Consolidates the repos file generation and build Earthfiles
> 
>     * Removes a bunch of the build logic around branches of this repo
> 
>     * Moves the spaceros install to /opt/spaceros from the created users home directory
>       
>       * I think it makes much more sense to put it there, as it's in line with the general ros approach
> 
>     * Includes some minor optimizations for the build
> 
>     * General cleanup of the documentation
> 
> 
> In searching through the overlaid applications and demonstrations in `docker`, I haven't found functional references to the`spaceros` install in the user's home directory. Though there are some docs references that will have to be updated alongside of this.
> 
> This does not yet separate dev/base targets, which is next on the todo list.

I am trying to reconcile all insues and PRs as of now...

this PR looks like closing #177, not #176? 

in fact it looks like it closes everything from #177 apart from #176 :)

**Comment by eholum** [comment]
> this PR looks like closing https://github.com/space-ros/space-ros/issues/177, not https://github.com/space-ros/space-ros/issues/176?
> in fact it looks like it closes everything from https://github.com/space-ros/space-ros/issues/177 apart from https://github.com/space-ros/space-ros/issues/176 :)

It certainly moves us towards the goal but #176 is correct. There are a few remaining tasks from #177 that will not get closed out, like setting naming conventions and actually getting the blessing from OSRF to push a few more images to the container registry.

This change sets us up to to a few of those things! Does that make sense?

**Comment by asimonov** [comment]
`ROS 2 repos check` shows a warning for ros2.repos that it starts with "---"

this is an artefact from yaml.dump at the end of merge-repos.py because technically yaml should start with document separator of "---"

shall we hide this check or fix yaml saving?

**Comment by eholum** [comment]
Yeah I assume that is from the yaml dump, but is unrelated to this PR so I didn't fix it. That said, I don't think it would be hard to so I opened: https://github.com/space-ros/space-ros/issues/180 to make sure we address it going forward.

**Comment by asimonov** [comment]
yes, i have seen it in history with '---'

**Comment by asimonov** [comment]
I finished my self-education exercise of understanding how space-ros docker is built, based on this PR.

It builds for me on Ubuntu.

And looking through the code before and after I think it retains the functionality.

**Comment by asimonov** [comment]
I have also built it on mac/arm64.

There is some warning during apt commands about using 'sudo' which I think comes from permissions of mounted caches. I suggest we change permissions (`mode=0777`) on those:

`
  RUN --mount=type=cache,mode=0777,target=/var/cache/apt,sharing=locked,id=cache_apt_cache \
      --mount=type=cache,mode=0777,target=/var/lib/apt,sharing=locked,id=lib_apt_cache \
`

**Comment by asimonov** [comment]
not sure if this is due to recent changes or it did not work before, but target  `earthly +repos-file` throws an error:

`
         +repos-file *failed* | The environment variable 'ROS_PACKAGE_PATH' must be set when using 'RPP'
         +repos-file *failed* | ERROR Earthfile:119:2
         +repos-file *failed* |       The command
         +repos-file *failed* |           RUN --no-cache sh scripts/generate-repos.sh --outfile ros2.repos --packages spaceros-pkgs.txt --excluded-packages excluded-pkgs.txt --rosdistro ${ROS_DISTRO}
         +repos-file *failed* |       did not complete successfully. Exit code 1
`

it is not used inside Earthfile, but mentioned in the USAGE.md

File: docs/USAGE.md
Comment (eholum) [reply docs/USAGE.md:55]: Good point. Added a note about it.

File: docs/USAGE.md
Comment (eholum) [reply docs/USAGE.md:22]: Gosh this is a weird one... I could swear this was working before. Also, I have no idea how this was working before since the old generate repos script did the same thing with an empty `AMENT_PREFIX_PATH`. I did note that the older images have `1.22` installed, whereas now I'm installing `1.23`. However, this check has been there for years:

https://github.com/ros-infrastructure/rosinstall_generator/blob/master/src/rosinstall_generator/generator.py#L203

In any case, this *should* be set once we actually have built and sourced anything for ROS. However, in this first instance `AMENT_PREFIX_PATH` is blank, so `ROS_PACKAGE_PATH` is blank, so it barfs. To get around this I just set it to the [empty spaceros directory](https://github.com/space-ros/space-ros/pull/170/files#diff-abd0a56959a8f57df0c5789b1d884f8fa1f0856228e4e0d60bb528831fede990R117). @xfiderek does that change look ok to you?

This should work now.

**Comment by asimonov** [comment]
The new changes work on my setup.

Do we have a plan to merge?

**Comment by eholum** [comment]
> The new changes work on my setup.
 > Do we have a plan to merge?

Thanks for evaluating! I don't think anything else is needed in this guy so if we want to go ahead and merge someone just needs to press the button. There is less of a rush considering we will have another few months before the next release to adjust this base build as needed.

File: docs/USAGE.md
Comment (eholum) [reply docs/USAGE.md:22]: Bump: I would appreciate another set of eyes just to make sure this isn't missing something silly.

Comment (xfiderek) [reply docs/USAGE.md:22]: I think the initial idea that I had was that we would run `generate-repos.sh` with `--exclude-installed false`. This argument defaults to true, because we would like this set for own downstream images and it is more convenient for users, but in base image we need to set it to false. I should have specified it more transparently, so that is for me. When we run it with this argument set to false, it should run smoothly and the modification to `export ROS_PACKAGE_PATH` line would not be required anymore.

**Comment by asimonov** [comment]
> > The new changes work on my setup.
> > Do we have a plan to merge?
> 
> Thanks for evaluating! I don't think anything else is needed in this guy so if we want to go ahead and merge someone just needs to press the button. There is less of a rush considering we will have another few months before the next release to adjust this base build as needed.

well, i kind want to start building on top.

question - if i build on top as separate branch and then this merge goes with a squash - will my commits still apply ok?

File: docs/USAGE.md
Comment (eholum) [reply docs/USAGE.md:22]: No worries! Ok that makes much more sense. I removed that commit and modified the build target in the Earthly file to set `--exclude-installed false`.

**Comment by eholum** [comment]
> well, i kind want to start building on top.

Me too! @asimonov @xfiderek if this looks good to you I think go ahead and slap an approval on it and we can get this merged.

> question - if i build on top as separate branch and then this merge goes with a squash - will my commits still apply ok?

Also yep. You may have to cherry-pick your changes on top of main instead of rebasing... but since the changes will be the same all the commits should still be just fine.

**Comment by asimonov** [comment]
> > well, i kind want to start building on top.
> 
> Me too! @asimonov @xfiderek if this looks good to you I think go ahead and slap an approval on it and we can get this merged.

Oh, can I approve? I doubt.. but once CI finishes on latest change let me try...

> 
> > question - if i build on top as separate branch and then this merge goes with a squash - will my commits still apply ok?
> 

> Also yep. You may have to cherry-pick your changes on top of main instead of rebasing... but since the changes will be the same all the commits should still be just fine.

well, I do not think I am that great with juggling commits between branches, I prefer to keep things simple. may be having smaller PRs with faster merges is a good idea...

**Comment by asimonov** [comment]
I certainly dont have 'approve' button

I think @Bckempa needs to review and approve, as requested

File: Earthfile
Comment (xfiderek) [inline Earthfile:129]: Just so I know - is the problem that I raised in here https://github.com/eholum/space-ros/pull/1#discussion_r1681445294 fixed?

File: Earthfile
Comment (xfiderek) [inline Earthfile:142]: Just so I know - is the problem that I raised in here https://github.com/eholum/space-ros/pull/1#discussion_r1681445294 fixed?

Comment (eholum) [reply Earthfile]: Ah, thanks @xfiderek.

I re-added the cache breaking echo from before. As an alternative, we could do something like https://github.com/eholum/space-ros/commit/7318ba4b798e91cdb445bb741d3befee0b46c9ac#diff-2705f6016a64189abd14041e2f3e8dbb829553295d9088e4eb42eccb5e61bf47R135 which will let us choose when to cache or not cache this layer.

That said, this isn't the bottleneck in terms of speed, but if rebuilding over and over locally it might be nice to go with that approach because we can skip duplicating caches. What do you think?

Comment (asimonov) [reply Earthfile]: i left a comment on original issue.

summary here: I favour using default caching behaviour of earthly/docker (as this is what most people expect) and documenting the aspects of cache growth/not picking up external (to dockerfile) changes by clearing the cache.

File: Earthfile
Comment (asimonov) [inline Earthfile:139]: question: will this result in non-caching of everyting happing in the build after this point? i.e. will 95%+ of the build steps have to be re-run?

File: Earthfile
Comment (eholum) [reply Earthfile]: +1. I think we can leave it as is with the `--no-cache` option. As Ezra said, in the future we shouldn't really be cloning during builds anyway, but we can leave this as is and acknowledge that we will have large caches.

File: Earthfile
Comment (eholum) [reply Earthfile:139]: Because we're only copying artifacts out of this layer, the cache *shouldn't change for downstream targets unless the actual artifacts have changed.

**Comment by franklinselva** [comment]
Do we know when this PR will be merged?

**Review by EzraBrooks — APPROVED** [review]
Approved per my comments in the meeting (which I believe were already captured by Erik in the comments here)

**Pull request merged**
